module/data/actor/templates/attributes.mjs

```diff
@@ -107,15 +107,26 @@ export default class AttributesFields {
     const statuses = this.parent.statuses;
     const noMovement = this.parent.hasConditionEffect("noMovement");
     const halfMovement = this.parent.hasConditionEffect("halfMovement");
-    const reduction = statuses.has("heavilyEncumbered")
-      ? CONFIG.DND5E.encumbrance.speedReduction.heavilyEncumbered
-      : statuses.has("encumbered") ? CONFIG.DND5E.encumbrance.speedReduction.encumbered : 0;
+    const encumbered = statuses.has("encumbered");
+    const heavilyEncumbered = statuses.has("heavilyEncumbered");
+    const exceedingCarryingCapacity = statuses.has("exceedingCarryingCapacity");
     const crawl = this.parent.hasConditionEffect("crawl");
-    Object.keys(CONFIG.DND5E.movementTypes).forEach(k => {
-      if ( reduction ) this.attributes.movement[k] = Math.max(0, this.attributes.movement[k] - reduction);
-      if ( (crawl && (k !== "walk")) || noMovement ) this.attributes.movement[k] = 0;
-      else if ( statuses.has("exceedingCarryingCapacity") ) this.attributes.movement[k] = 5;
-      else if ( halfMovement ) this.attributes.movement[k] = Math.floor(this.attributes.movement[k] * 0.5);
-    });
+    const units = this.attributes.movement.units;
+    for ( const type in CONFIG.DND5E.movementTypes ) {
+      let speed = this.attributes.movement[type];
+      if ( noMovement || (crawl && (type !== "walk")) ) speed = 0;
+      else {
+        if ( halfMovement ) speed *= 0.5;
```

Comment:
Applying halving before the flat reduction. I don't think there's actually a rule for the order to resolve exhaustion 2 and encumbrance and other flat speed reductions. I think this case usually the effect that is applied first is calculated first. But I don't think that makes sense here, because dropping all your items to remove the encumbrance status and then picking them back up shouldn't change your speed. Anyway, I think the encumbrance reduction should be applied after exhaustion 2, but opinions might differ here.

Another problem I see here handling the movement speed calculation here is that they are always applied after Active Effects, which will surely lead to incorrect speed calculation if for example slow effects from spells are applied to the actor, which should probably be applied after encumbrance and exhaustion 2.

Reply:
Yeah this is an area where the rules do not give clear guidance as far as I know. I don't think we really have a good way to apply effects in the order they are received, and implementing a way of tracking that seems like it would be far more effort than it's worth. So we should just pick an order (at least for now), and this proposed order seems reasonable enough to me.